File: app/src/main/java/com/chess/cryptobot/content/ContextHolder.java
package com.chess.cryptobot.content;

import android.content.Context;

import androidx.fragment.app.Fragment;

import com.chess.cryptobot.exceptions.ItemNotFoundException;
import com.chess.cryptobot.model.ViewItem;
import com.chess.cryptobot.view.MainActivity;
import com.chess.cryptobot.view.MainFragment;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public abstract class ContextHolder {
    private final Fragment fr;
    private List<ViewItem> viewItems;
    private final Preferences prefs;

    protected ContextHolder(Fragment fr) {
        this.fr = fr;
        initFields();
        prefs = initPrefs(fr.getContext());
        Set<String> itemsSet = prefs.getItems();
        initViewItems(itemsSet);
    }

    public void initFields() {
        viewItems = new ArrayList<>();
    }

    protected abstract Preferences initPrefs(Context context);

    protected abstract void initViewItems(Set<String> itemsSet);

    public synchronized void add(ViewItem viewItem) {
        addItemToList(viewItem);
        getMainFragment().addItem();
        Preferences preferences = getPrefs();
        preferences.addItem(viewItem.getName());
    }

    protected synchronized void addItemToList(ViewItem viewItem) {
        if (viewItems.contains(viewItem)) return;
        viewItems.add(viewItem);
    }

    protected void retainAll(List<ViewItem> viewItems) {
        List<ViewItem> invalidItems = new ArrayList<>(getViewItems());
        getViewItems().retainAll(viewItems);

        invalidItems.forEach(viewItem -> {
            if (!getViewItems().contains(viewItem)) {
                removeFromPrefs(viewItem);
            }
        });
        getMainFragment().updateAllItems();
    }

    private void removeFromAdapter(ViewItem viewItem) {
        getMainFragment().deleteItemByPosition(viewItems.indexOf(viewItem));
    }

    private void removeFromPrefs(ViewItem viewItem) {
        getPrefs().removeItem(viewItem.getName());
    }

    public synchronized void remove(ViewItem item) {
        if (item == null) return;
        removeFromAdapter(item);
        viewItems.remove(item);
        removeFromPrefs(item);
    }

    public synchronized void setItem(ViewItem updatedItem) {
        for (ViewItem item : viewItems) {
            if (item.equals(updatedItem)) {
                viewItems.set(viewItems.indexOf(item), updatedItem);
                break;
            }
        }

        getMainFragment().updateItem(updatedItem);
    }

    public void updateAllItems() {
        viewItems.forEach(this::updateItem);
    }

    protected abstract void updateItem(ViewItem item);

    protected ViewItem getItemByName(String itemName) throws ItemNotFoundException {
        for (ViewItem item : viewItems) {
            if (item.getName().equals(itemName)) return item;
        }
        throw new ItemNotFoundException(itemName);
    }

    public List<ViewItem> getViewItems() {
        return this.viewItems;
    }

    public Context getContext() {
        return this.fr.getContext();
    }

    public Preferences getPrefs() {
        return this.prefs;
    }

    protected MainFragment getMainFragment() {
        return (MainFragment) this.fr;
    }

    protected MainActivity getMainActivity() {
        return (MainActivity) this.fr.getActivity();
    }

    public void makeToast(String message) {
        this.getMainFragment().makeToast(message);
    }

    public void showSpinner() {
        this.getMainFragment().showSpinner();
    }

    public void hideSpinner() {
        this.getMainFragment().hideSpinner();
    }
}

File: app/src/main/java/com/chess/cryptobot/content/ContextHolder.kt
package com.chess.cryptobot.content

import android.content.Context
import androidx.fragment.app.Fragment
import com.chess.cryptobot.exceptions.ItemNotFoundException
import com.chess.cryptobot.model.ViewItem
import com.chess.cryptobot.view.MainFragment

abstract class ContextHolder protected constructor(private val fr: Fragment) {
    var viewItems: MutableList<ViewItem> = ArrayList()
    val context: Context = fr.requireContext()

    lateinit var prefs: Preferences

    protected val mainFragment: MainFragment<*>
        get() = this.fr as MainFragment<*>

    open fun init(): ContextHolder {
        initViewItems(initPrefs(context).items)
        return this
    }

    protected abstract fun initPrefs(context: Context): Preferences

    protected abstract fun initViewItems(itemNamesSet: HashSet<String>?): MutableList<ViewItem>

    open fun add(viewItem: ViewItem) {
        addItemToList(viewItem)
        mainFragment.addItem()
        val preferences = prefs
        preferences.addItem(viewItem.name)
    }

    protected fun addItemToList(viewItem: ViewItem) {
        if (viewItems.contains(viewItem)) return
        viewItems.add(viewItem)
    }

    protected fun retainAll(viewItems: List<ViewItem>) {
        val invalidItems = ArrayList(this.viewItems)
        this.viewItems.retainAll(viewItems)

        invalidItems.forEach { viewItem ->
            if (!this.viewItems.contains(viewItem)) {
                removeFromPrefs(viewItem)
            }
        }
        mainFragment.updateAllItems()
    }

    private fun removeFromAdapter(viewItem: ViewItem) {
        mainFragment.deleteItemByPosition(viewItems.indexOf(viewItem))
    }

    private fun removeFromPrefs(viewItem: ViewItem) {
        prefs.removeItem(viewItem.name)
    }

    fun remove(item: ViewItem?) {
        if (item == null) return
        removeFromAdapter(item)
        viewItems.remove(item)
        removeFromPrefs(item)
    }

    fun setItem(updatedItem: ViewItem?) {
        for (item in viewItems) {
            if (item == updatedItem) {
                viewItems[viewItems.indexOf(item)] = updatedItem
                break
            }
        }

        mainFragment.updateItem(updatedItem)
    }

    open fun updateAllItems() {
        viewItems.forEach { viewItem -> this.updateItem(viewItem) }
    }

    protected abstract fun updateItem(item: ViewItem)

    @Throws(ItemNotFoundException::class)
    protected fun getItemByName(itemName: String): ViewItem {
        for (item in viewItems) {
            if (item.name == itemName) return item
        }
        throw ItemNotFoundException(itemName)
    }

    fun makeToast(message: String) {
        this.mainFragment.makeToast(message)
    }

    fun showSpinner() {
        this.mainFragment.showSpinner()
    }

    fun hideSpinner() {
        this.mainFragment.hideSpinner()
    }
}